// react-instantsearch-core/dist/cjs/core/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IndexProvider = exports.IndexConsumer = exports.InstantSearchProvider = exports.InstantSearchConsumer = void 0;

var _react = require("react");

var _createContext = (0, _react.createContext)({
  onInternalStateUpdate: function onInternalStateUpdate() {
    return undefined;
  },
  createHrefForState: function createHrefForState() {
    return '#';
  },
  onSearchForFacetValues: function onSearchForFacetValues() {
    return undefined;
  },
  onSearchStateChange: function onSearchStateChange() {
    return undefined;
  },
  onSearchParameters: function onSearchParameters() {
    return undefined;
  },
  store: {},
  widgetsManager: {},
  mainTargetedIndex: ''
}),
    InstantSearchConsumer = _createContext.Consumer,
    InstantSearchProvider = _createContext.Provider;

exports.InstantSearchProvider = InstantSearchProvider;
exports.InstantSearchConsumer = InstantSearchConsumer;

var _createContext2 = (0, _react.createContext)(undefined),
    IndexConsumer = _createContext2.Consumer,
    IndexProvider = _createContext2.Provider;

exports.IndexProvider = IndexProvider;
exports.IndexConsumer = IndexConsumer;
